File: packages/sea-core-ios/Sources/SEACore/SEAError.swift
import Foundation

/// SEA's error taxonomy (contract §3.5, spec §7.2).
public enum SEAError: Error, Equatable {
    case network(underlying: String)
    case timeout
    case cancelled
    case invalidAuthorizeURL(reason: SEAInvalidURLReason)
    case serverError(statusCode: Int)

    /// §10.4: the WebAuthn fallback path itself could not be started. This
    /// is *not* thrown just because the embedded WKWebView ceremony is
    /// unsupported or fails live — that case is handled transparently by
    /// routing the whole login attempt through `SEAFallbackAuthRunner`
    /// (`ASWebAuthenticationSession`), which still resolves via the normal
    /// `onCaptured`/`onCancelled`/`onError` contract. This case is reachable
    /// only when the fallback itself cannot run, e.g.
    /// `SEAEnvironment.current.callbackScheme` is empty (fail-closed host
    /// misconfiguration, contract §3.3) — there is no scheme an
    /// `ASWebAuthenticationSession` could ever be started with.
    case webauthnUnavailable

    /// Phase 2 (§21). Kill-switch fallback is a Bankerise-SDK-level concern
    /// (gateway-driven `authMode`), not something SEA itself decides. This
    /// case is declared for taxonomy stability but is never thrown by this
    /// package in Phase 1.
    case killSwitched
}

/// Reasons an authorize URL fails validation (contract §3.5, §5).
///
/// Conforms to `Error` (not shown in the contract's code block) because
/// `SEAAuthorizeURLValidator.validate` returns `Result<URL, SEAInvalidURLReason>`
/// (contract §5), and `Result`'s failure type must conform to `Error`.
public enum SEAInvalidURLReason: String, Error {
    case scheme
    case host
    case userinfo
    case port
    case length
    case malformed
}

File: packages/sea-core-android/src/main/java/com/bankerise/sea/core/SEAError.kt
package com.bankerise.sea.core

/**
 * SEA's error taxonomy (contract §3.5, spec §7.2).
 */
sealed class SEAError {
    data class Network(val underlying: String) : SEAError()
    object Timeout : SEAError()
    object Cancelled : SEAError()
    data class InvalidAuthorizeUrl(val reason: InvalidUrlReason) : SEAError()
    data class ServerError(val statusCode: Int) : SEAError()

    /** Phase 2 (§10). Declared for taxonomy stability; never thrown in Phase 1. */
    object WebauthnUnavailable : SEAError()

    /** Phase 2 (§21). Declared for taxonomy stability; never thrown in Phase 1. */
    object KillSwitched : SEAError()

    override fun equals(other: Any?): Boolean = when {
        this === other -> true
        other !is SEAError -> false
        else -> when (this) {
            is Network -> other is Network && underlying == other.underlying
            Timeout -> other is Timeout
            Cancelled -> other is Cancelled
            is InvalidAuthorizeUrl -> other is InvalidAuthorizeUrl && reason == other.reason
            is ServerError -> other is ServerError && statusCode == other.statusCode
            WebauthnUnavailable -> other is WebauthnUnavailable
            KillSwitched -> other is KillSwitched
        }
    }

    override fun hashCode(): Int = when (this) {
        is Network -> underlying.hashCode()
        Timeout -> 0
        Cancelled -> 1
        is InvalidAuthorizeUrl -> reason.hashCode()
        is ServerError -> statusCode
        WebauthnUnavailable -> 10
        KillSwitched -> 11
    }
}

/**
 * Reasons an authorize URL fails validation (contract §3.5, §5).
 */
enum class InvalidUrlReason {
    SCHEME,
    HOST,
    USERINFO,
    PORT,
    LENGTH,
    MALFORMED
}

class InvalidUrlException(val reason: InvalidUrlReason) : Exception(reason.name)
